### DISCUSSION TITLE
Trying to load an onnx file in C

### DISCUSSION THREAD

**Discussion by victorbrandaoa** [microsoft/onnxruntime#21265] [API Q&A]
I tried to use the onnxruntime_c_api.h in C, but I couldn't. I intend to train a model in python and save it as onnx, then load it in C to do inference. Is there any documentation to help me with that? Bear in mind that I am using Linux.

**Comment by manickavela29** [comment]
Hi @victorbrandaoa,

Is your question towards asking how to include the headerfile into your project? or
how to create a c program to handle the inference.

this example would be of some help with program [onnxrt c example](https://github.com/microsoft/onnxruntime-inference-examples/blob/main/c_cxx/fns_candy_style_transfer/fns_candy_style_transfer.c.)

I would suggest to try with c++ than c,
for ease of use and for community support.

**Reply by victorbrandaoa** [reply]
In my context problem, I need to create a C program (only in C). I want to create a C program to handle the inference, but I am not sure how I can install onnxruntine to do so. The way I installed I only got `.h` files, but they represent interfaces, so where are the `.c` files with the implementation? Do you have any documentation about it? Thanks

**Reply by Craigacp** [reply]
You probably want to depend on the shared library rather than compiling onnx runtime into your project. The example C code has instructions for Windows, but it's pretty similar on Linux if you swap out bat for sh and then it depends what your C build system is - https://github.com/microsoft/onnxruntime-inference-examples/blob/main/c_cxx/README.md.

**Reply by manickavela29** [reply]
yes, I agree on the same,
which build system are you using @victorbrandaoa?
If you are using Makefile directly then

- Install onnxruntime (either system directly or any environment)
- Compiling -> Ensure that header files are included(onnxruntime_c_api.h, etc)
- Linking      -> Ensure that onnxruntime lib paths are added to it 
I just asked chatgpt and got the below, I hope you can adapt to your usecase ,

```

# Makefile for building a C++ application with ONNX Runtime

# Define compiler and flags
CXX = g++
CXXFLAGS = -O2 -std=c++14

# Define ONNX Runtime version and installation paths
ONNXRUNTIME_VERSION = 1.17.1
ONNXRUNTIME_URL = https://github.com/microsoft/onnxruntime/releases/download/v$(ONNXRUNTIME_VERSION)/onnxruntime-linux-x64-$(ONNXRUNTIME_VERSION).tgz
ONNXRUNTIME_DIR = onnxruntime-linux-x64-$(ONNXRUNTIME_VERSION)
ONNXRUNTIME_LIB = $(ONNXRUNTIME_DIR)/lib
ONNXRUNTIME_INCLUDE = $(ONNXRUNTIME_DIR)/include

# Define target executable
TARGET = my_application

# Define source files
SRC = main.cpp

# Define object files
OBJ = $(SRC:.cpp=.o)

# Rules
all: $(TARGET)
# Download and extract ONNX Runtime
$(ONNXRUNTIME_DIR):
	wget $(ONNXRUNTIME_URL)
	tar -xzf onnxruntime-linux-x64-$(ONNXRUNTIME_VERSION).tgz

# Compile object files
%.o: %.cpp $(ONNXRUNTIME_DIR)
	$(CXX) $(CXXFLAGS) -I$(ONNXRUNTIME_INCLUDE) -c $< -o $@

# Link object files to create the executable
$(TARGET): $(OBJ) $(ONNXRUNTIME_DIR)
	$(CXX) $(CXXFLAGS) -o $(TARGET) $(OBJ) -L$(ONNXRUNTIME_LIB) -lonnxruntime

# Clean up
clean:
	rm -f $(TARGET) $(OBJ)
	rm -rf $(ONNXRUNTIME_DIR)
	rm -f onnxruntime-linux-x64-$(ONNXRUNTIME_VERSION).tgz

.PHONY: all clean
```

But since you are writing a  c code, you have to be extra careful.

**Reply by victorbrandaoa** [reply]
Is running the following command enough? When I ran the command, can I just create a main.c in the root directory of the onnxruntime, just to test?

```sh
./build.sh --config RelWithDebInfo --build_shared_lib --parallel --compile_no_warning_as_error --skip_submodule_sync
```

How can I ensure that onnxruntime lib paths are added to it?

The error I'm facing is:

```sh
main.c:(.text+0x954): undefined reference to `OrtGetApiBase'
```

**Reply by manickavela29** [reply]
That would be a very bad way to go about it. (yes, it can be done somehow but no)
I would suggest you to read more about C/C++ project, build system (make/cmake), linking necessary libraries and etc.
Its a long journey ahead with this.

Fundamentally you need to understand, how to create a c/c++ project and link onnxruntime lib with it.
